### ISSUE TITLE
tensor.triu_(1) not working properly with large matrix

### ISSUE THREAD

**Issue by ThisisBillhe** [pytorch/pytorch#136611]
### 🐛 Describe the bug

Ideally, tensor.triu_(1) should fill lower triangular part with 0. However, it fails to do so when the matrix is large. For example:
```
q_len = 100000
causal_mask = torch.full((q_len, q_len), float('-inf')).to(device='cuda')
causal_mask.triu_(1)  # Fill lower triangular part with 0
```
result:
```
tensor([[0., -inf, -inf,  ..., -inf, -inf, -inf],
        [0., 0., -inf,  ..., -inf, -inf, -inf],
        [0., 0., 0.,  ..., -inf, -inf, -inf],
        ...,
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf],
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf],
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf]], device='cuda:0')
```
The first row that is not started with 0 is 42950.

### Versions

Collecting environment information...
PyTorch version: 2.1.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.29.6
Libc version: glibc-2.35

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.19.91-014.15-kangaroo.alios7.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 535.54.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU: skipped

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] open_clip_torch==2.26.1
[pip3] torch==2.1.2
[pip3] torchvision==0.16.2
[pip3] triton==2.1.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] open-clip-torch           2.26.1                   pypi_0    pypi
[conda] torch                     2.1.2                    pypi_0    pypi
[conda] torchvision               0.16.2                   pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

**Comment by d-kleine** [comment]
Please test it with the latest stable release (2.4.1) of pytorch. I can only test this with `'cpu', and it seems to work correct

```python
import torch

q_len = 100000
causal_mask = torch.full((q_len, q_len), float('-inf')).to(device='cpu')
causal_mask.triu_(1)  # Fill lower triangular part with 0
```


```
tensor([[0., -inf, -inf,  ..., -inf, -inf, -inf],
        [0., 0., -inf,  ..., -inf, -inf, -inf],
        [0., 0., 0.,  ..., -inf, -inf, -inf],
        ...,
        [0., 0., 0.,  ..., 0., -inf, -inf],
        [0., 0., 0.,  ..., 0., 0., -inf],
        [0., 0., 0.,  ..., 0., 0., 0.]])
```

Version:
```python
torch.__version__
```

```
'2.4.1+cu121'
```

**Comment by ThisisBillhe** [comment]
Hi, I test it with torch.__version__='2.4.1+cu121', on a single A100 GPU. The bug is still there. The first row that is not started with 0 is still 42950.

```
causal_mask
tensor([[0., -inf, -inf,  ..., -inf, -inf, -inf],
        [0., 0., -inf,  ..., -inf, -inf, -inf],
        [0., 0., 0.,  ..., -inf, -inf, -inf],
        ...,
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf],
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf],
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf]], device='cuda:0')
```
On CPU, it is fine.

**Comment by mikaylagawarecki** [comment]
I can repro

**Comment by jamarju** [comment]
Another minimal reproduce on torch '2.4.1+cu121'

This works:

```py
x = torch.randn(1<<16, 1<<17, device="cuda", dtype=torch.float16)
assert torch.allclose(torch.tril(x).cpu(), torch.tril(x.cpu()))
```

But this triggers the assertion.

```py
x = torch.randn(1<<16, 1<<17, device="cuda", dtype=torch.float16)
assert torch.allclose(torch.tril(x).cpu(), torch.tril(x.cpu()))
```

Looks like an uint32 index overflow issue.

**Comment by nikitaved** [comment]
I could have a look unless someone is working on it.

**Comment by Wangmerlyn** [comment]
I can reproduce this issue also, it seems the problem is with cuda.
I've tried setting the original matrix `dtype=torch.bfloat16` and also `dtype=torch.float32`, both dtypes have this issue on a cuda device.
 `triu(1)` works normally on cpu.

**Comment by Wangmerlyn** [comment]
The interesting part about this bug is that it works different for `torch.triu()` and `torch.Tensor.triu_()`.

For example when I'm trying to take the upper triangular of a mask that is full of `-inf` with `torch.Tensor.triu_()`
``` python
causal_mask = torch.full((128*1024, 128*1024), fill_value=torch.finfo(torch.bfloat16).min, dtype=torch.bfloat16, device="cuda")
causal_mask.triu_(1)
```
I get the same output as 
``` python
tensor([[0., -inf, -inf,  ..., -inf, -inf, -inf],
        [0., 0., -inf,  ..., -inf, -inf, -inf],
        [0., 0., 0.,  ..., -inf, -inf, -inf],
        ...,
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf],
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf],
        [-inf, -inf, -inf,  ..., -inf, -inf, -inf]], device='cuda:0')
```
![image](https://github.com/user-attachments/assets/b5b16d4d-da04-4838-a35a-020d91b4d3f7)
The point where this triangularized matrix went wrong is around 33k for a matrix size of (128*1024, 128*1024)
But if I use `torch.triu()`
``` python 
causal_mask = torch.full((128*1024, 128*1024), fill_value=torch.finfo(torch.bfloat16).min, dtype=torch.bfloat16, device="cuda")
causal_mask. = torch.triu(causal_mask, diagonal=1)
```
I get the output wrong in a different way.
![image](https://github.com/user-attachments/assets/2811f30a-0fa2-4407-993c-71d357235290)

**Comment by jamarju** [comment]
Here is a workaround in Triton in case someone is interested. 

Note 1: `int64` types are required for the offsets computation to be able to address matrices larger than 2^32 elements which is likely the problem in torch's CUDA implementation.

Note 2: this kernel can operate safely on matrices > 2^32 elements but has a `65535 * BLOCK_SIZE` limit on either dimension as it appears 2D grid indices are internally represented as uint16 by Triton. Anyhow hope this helps.

```py
import triton
import triton.language as tl

@triton.jit
def tril_kernel(
    input_ptr, output_ptr, stride_m, stride_n, M, N, diagonal, BLOCK_SIZE: tl.constexpr
):
    pid_m = tl.cast(tl.program_id(0), tl.int64)  # Row block ID
    pid_n = tl.cast(tl.program_id(1), tl.int64)  # Column block ID

    # Compute row and column indices for this block
    row_indices = pid_m * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
    col_indices = pid_n * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)

    # Create a mask for valid rows and columns within matrix bounds
    row_mask = row_indices < M
    col_mask = col_indices < N

    # Load input block
    input_block = tl.load(
        input_ptr + row_indices[:, None] * stride_m + col_indices[None, :] * stride_n,
        mask=row_mask[:, None] & col_mask[None, :],
        other=0.0,
    )

    # Mask out elements above the specified diagonal
    tril_mask = row_indices[:, None] >= (col_indices[None, :] - diagonal)
    output_block = tl.where(tril_mask, input_block, 0.0)

    # Store output block
    tl.store(
        output_ptr + row_indices[:, None] * stride_m + col_indices[None, :] * stride_n,
        output_block,
        mask=row_mask[:, None] & col_mask[None, :],
    )

def tril_triton(input_tensor, diagonal=0):
    # Ensure input is a 2D matrix
    assert input_tensor.dim() == 2, "Input must be a 2D tensor"

    # Get matrix dimensions
    M, N = input_tensor.shape

    # Allocate output tensor
    output_tensor = torch.empty_like(input_tensor)

    # Define block size (tune for your GPU hardware)
    BLOCK_SIZE = 128

    # Launch the Triton kernel with a 2D grid (for row and column blocks)
    # WARNING: Triton has a max grid size of 65535 on either dimension
    grid = (triton.cdiv(M, BLOCK_SIZE), triton.cdiv(N, BLOCK_SIZE))
    print(grid)
    tril_kernel[grid](
        input_tensor,
        output_tensor,
        input_tensor.stride(0),
        input_tensor.stride(1),
        M,
        N,
        diagonal,
        BLOCK_SIZE,
    )

    return output_tensor
```

**Comment by Wangmerlyn** [comment]
> Here is a workaround in Triton in case someone is interested.
> 
> Note 1: `int64` types are required for the offsets computation to be able to address matrices larger than 2^32 elements which is likely the problem in torch's CUDA implementation.
> 
> Note 2: this kernel can operate safely on matrices > 2^32 elements but has a `65535 * BLOCK_SIZE` limit on either dimension as it appears 2D grid indices are internally represented as uint16 by Triton. Anyhow hope this helps.
> 
> ```python
> import triton
> import triton.language as tl
> 
> @triton.jit
> def tril_kernel(
>     input_ptr, output_ptr, stride_m, stride_n, M, N, diagonal, BLOCK_SIZE: tl.constexpr
> ):
>     pid_m = tl.cast(tl.program_id(0), tl.int64)  # Row block ID
>     pid_n = tl.cast(tl.program_id(1), tl.int64)  # Column block ID
> 
>     # Compute row and column indices for this block
>     row_indices = pid_m * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
>     col_indices = pid_n * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
> 
>     # Create a mask for valid rows and columns within matrix bounds
>     row_mask = row_indices < M
>     col_mask = col_indices < N
> 
>     # Load input block
>     input_block = tl.load(
>         input_ptr + row_indices[:, None] * stride_m + col_indices[None, :] * stride_n,
>         mask=row_mask[:, None] & col_mask[None, :],
>         other=0.0,
>     )
> 
>     # Mask out elements above the specified diagonal
>     tril_mask = row_indices[:, None] >= (col_indices[None, :] - diagonal)
>     output_block = tl.where(tril_mask, input_block, 0.0)
> 
>     # Store output block
>     tl.store(
>         output_ptr + row_indices[:, None] * stride_m + col_indices[None, :] * stride_n,
>         output_block,
>         mask=row_mask[:, None] & col_mask[None, :],
>     )
> 
> def tril_triton(input_tensor, diagonal=0):
>     # Ensure input is a 2D matrix
>     assert input_tensor.dim() == 2, "Input must be a 2D tensor"
> 
>     # Get matrix dimensions
>     M, N = input_tensor.shape
> 
>     # Allocate output tensor
>     output_tensor = torch.empty_like(input_tensor)
> 
>     # Define block size (tune for your GPU hardware)
>     BLOCK_SIZE = 128
> 
>     # Launch the Triton kernel with a 2D grid (for row and column blocks)
>     # WARNING: Triton has a max grid size of 65535 on either dimension
>     grid = (triton.cdiv(M, BLOCK_SIZE), triton.cdiv(N, BLOCK_SIZE))
>     print(grid)
>     tril_kernel[grid](
>         input_tensor,
>         output_tensor,
>         input_tensor.stride(0),
>         input_tensor.stride(1),
>         M,
>         N,
>         diagonal,
>         BLOCK_SIZE,
>     )
> 
>     return output_tensor
> ```

Thank you so much for this work around. My solution for my last code segment was initializing the causal mask and triangularize it on cpu since the problem is with cuda. Even though the matrix was as large as 128k*128k, initializing and triangularization won't take up as much time as merely seconds.

**Comment by grad0s** [comment]
see https://github.com/pytorch/pytorch/issues/136211
